docs/developer/logging-guidelines.md

```diff
@@ -0,0 +1,407 @@
+# Logging Guidelines
+
+This document describes the logging guidelines for the control plane of NGINX Kubernetes Gateway (NKG).
+
+> The data plane logging is not covered here: such a concern is owned by NGINX developers, and NKG developers
+> don't have control over it.
+
+## Table of Contents
+
+<!-- TOC -->
+
+- [Logging Guidelines](#logging-guidelines)
+  - [Table of Contents](#table-of-contents)
+  - [Requirements](#requirements)
+    - [User Stories](#user-stories)
+      - [Common Stories](#common-stories)
+      - [Stories For User](#stories-for-user)
+      - [Stories For Developers](#stories-for-developers)
+  - [Logging Library Choice](#logging-library-choice)
+  - [Guidelines](#guidelines)
+    - [How to Log](#how-to-log)
+      - [Log Levels](#log-levels)
+      - [Log Messages](#log-messages)
+      - [Context in Messages](#context-in-messages)
+      - [Examples of Log Messages](#examples-of-log-messages)
+      - [Message Guidelines](#message-guidelines)
+    - [Log Formatting](#log-formatting)
+    - [When to Log](#when-to-log)
+    - [What not to Log](#what-not-to-log)
+    - [Performance](#performance)
+    - [Logger Dependency in Code](#logger-dependency-in-code)
+      - [Logger Initialization](#logger-initialization)
+      - [Logger Injection](#logger-injection)
+      - [Special Case - Reconciler](#special-case---reconciler)
+      - [Unit Tests](#unit-tests)
+    - [External Libraries](#external-libraries)
+    - [Evolution](#evolution)
+  - [External Resources](#external-resources)
+
+<!-- TOC -->
+
+## Requirements
+
+Before delving into the guidelines, we must first establish the high-level requirements for logging to ensure a clear
+understanding of what these guidelines intend to address:
+
+> We don't anticipate these high-level requirements to change for the duration of the project. If changes occur,
```

Comment:
Not sure if this is necessary... If we update the guidelines, we update the guidelines. They have implied persistence because we took the time to write them down.